### PR TITLE
Add tests and workflow for package build / testing

### LINKED ISSUE #20: Add unit tests
While #17 added github workflows, right now this library does not have unit tests. This is blocked on `colcon-cargo` being a dependency but not being released yet in pip.
Once it is released (https://github.com/colcon/colcon-cargo/issues/22) we should add unit tests to make sure this library performs as intended

### PR DESCRIPTION
**Pull request by luca-della-vedova**
Based of #19, I can merge it there to have a single PR with functionality and tests or keep it separate and retarget `main` once #19 is merged.

Note that CI will be red until the upstream https://github.com/colcon/colcon-cargo/pull/39 is merged, since this PR adds a test for `colcon test`, which is not implemented right now.
Furthermore, it currently uses a proposed strategy for injecting setup commands into the colcon reusable workflow in https://github.com/colcon/ci/pull/33. If the PR is merged we should revert to main, if other strategies are proposed we should adopt them before merging this in.

### CODE REVIEW

**Comment by luca-della-vedova** [comment]
If no activity happens on the upstream `colcon/ci` PR I'll proceed and copy the action with the required change in this repo to make the workflow green

**Comment by luca-della-vedova** [comment]
As mentioned in https://github.com/colcon/colcon-ros-cargo/pull/22#issuecomment-2297799790 I proceeded  to copy the action here in https://github.com/colcon/colcon-ros-cargo/pull/22/commits/0abad0d384233c600c90690236c73bd9671ca8e4, we can revisit that if some feedback comes from the upstream reusable workflow maintainers

It seems `cargo-ament-build` doesn't deal with binaries on Windows properly which makes the CI fail there. I created a temporary branch and added it to the CI to prove that it changes it from red to green in https://github.com/colcon/colcon-ros-cargo/pull/22/commits/566e6f967ff913a86ede7b76fa068d7575541c54.

PR to merge the changes back to main and make sure this package works on Windows in https://github.com/ros2-rust/cargo-ament-build/pull/8.

Once the `cargo-ament-build` PR is merged (and `cargo-ament-build` released) I'll revert the custom branch and this should be good to go

File: test/rust-sample-package/Cargo.toml
Comment (mxgrey) [inline test/rust-sample-package/Cargo.toml]: Just curious, whose name is this and how did it land here? I feel hesitant to use the name of someone who's not involved in the project.

Comment (luca-della-vedova) [reply test/rust-sample-package/Cargo.toml]: I [copied](https://github.com/colcon/colcon-cargo/blob/main/test/rust-sample-package/Cargo.toml) the package from `colcon-cargo` hence kept the original attribution.
It is true that at the end of the day it's just an empty template where the only code contribution was made by me though (adding code that passes / fails test)

Comment (mxgrey) [reply test/rust-sample-package/Cargo.toml]: I see, it's an interesting dilemma. On one hand, I wouldn't want to use someone's contact info without their explicit consent. On the other hand, it could be interpreted as an act of attribution, as you're saying. 

Since we're using the same code in the same way that the original author intended it, just in a different (but strongly related) package, I suppose I don't object to keeping it as-is. But if anyone has a different perspective, I'd appreciate hearing it. 

I'll tag @bergercookie and maybe if he sees the notification, he can give a 👍/👎 on the usage.

Comment (esteve) [reply test/rust-sample-package/Cargo.toml]: Since this sample package is not meant to be published as is, I'd change the author and email to something non-specific (Test McTestface <notareal@email.address>) to avoid attribution, but also prevent @bergercookie getting spammed. I'd do the same for `package.xml`.

Comment (luca-della-vedova) [reply test/rust-sample-package/Cargo.toml]: https://github.com/colcon/colcon-ros-cargo/pull/22/commits/a62ff2221b0038b922d99702c9b2dd66bce91289

File: colcon_ros_cargo/task/ament_cargo/build.py
Comment (maspe36) [inline colcon_ros_cargo/task/ament_cargo/build.py]: So we delete the previous `config.toml`if it exists?

Nit: Perhaps unnecessary for folks who spend a lot of time in python, but I would suggest wrapping this in a small free function with a clear name.

`def _delete_file(f)` or something similar.

Also, we should probably explain why we pass on the exception 
`# No work for us to do, the file already doesn't exist` or something like that.

**Review by maspe36 — DISMISSED** [review]
LGTM

File: colcon_ros_cargo/task/ament_cargo/build.py
Comment (luca-della-vedova) [reply colcon_ros_cargo/task/ament_cargo/build.py]: After looking at this further I feel the step might not be necessary, unless there is some extreme corner case, opening a file as `w` by default already deletes its whole content and replaces it, so I removed the whole block in https://github.com/colcon/colcon-ros-cargo/pull/22/commits/0612fda1daa42202bdc2b89e30c75b1f2e11bdbf

The only case in which this would behave differently (I believe) is if users created a `.cargo/config.toml` that was a symlink to their own file, then ran a `colcon build`, previously the link would be deleted and a new config file would be created, now the symlink would be followed and edited.
I'd argue that it is such a corner case that I don't even know what the intended behavior would be

**Review by maspe36 — APPROVED** [review]

**Pull request merged**
